tests/test_phase_psf.py

```diff
@@ -985,6 +985,83 @@ def test_gc():
+    psfGSP = psf.withGSParams(galsim.GSParams(folding_threshold=6e-3))
+    # Don't worry about repr for DummyScreen
+    do_pickle(psf, irreprable=True)
+    do_pickle(psfGSP, irreprable=True)
+
+    # We can't use do_pickle with func that involves a random number generator, so just hard-code
+    # the equivalent here.
+    try:
+        import cPickle as pickle
+    except:
+        import pickle
+    import copy
+    copiers = [lambda x:pickle.loads(pickle.dumps(x)), copy.copy, copy.deepcopy]
+    drawKwargs = dict(nx=32, ny=32, scale=0.2, method='phot', n_photons=10000)
+
+    rng = galsim.BaseDeviate(57721)
+    img = psf.drawImage(rng=rng, **drawKwargs)
+    for copier in copiers:
+        psf2 = copier(psf)
+        assert psf2 is not psf
+        rng = galsim.BaseDeviate(57721) # reset rng
+        img2 = psf2.drawImage(rng=rng, **drawKwargs)
+        assert img2 == img
+
+    # Repeat but make a fresh (not previously drawn) source psf each time.  I.e.,
+    # make sure that drawing and copy/pickling are commutative
+    testimgs = []
+    for copier in copiers:
+        psf = psl.makePSF(exptime=0.02, time_step=0.01, diam=1.1, lam=1000.0)
+        psf2 = copier(psf)
+        assert psf2 is not psf
+        rng = galsim.BaseDeviate(57721) # reset rng
+        testimgs.append(psf2.drawImage(rng=rng, **drawKwargs))
+    # Now draw the fresh psf and compare
+    psf = psl.makePSF(exptime=0.02, time_step=0.01, diam=1.1, lam=1000.0)
+    rng = galsim.BaseDeviate(57721)
+    img = psf.drawImage(rng=rng, **drawKwargs)
+    for testimg in testimgs:
+        assert testimg == img
+
+    # Want to check that copying and withGSParams are also commutative.  So repeat steps above
+    # using psfGSP as the source/test.
+    rng = galsim.BaseDeviate(57721)
+    img = psfGSP.drawImage(rng=rng, **drawKwargs)  # uncopied withGSP img
+    for copier in copiers:
+        psf2 = copier(psfGSP)
+        assert psf2 is not psfGSP
+        rng = galsim.BaseDeviate(57721) # reset rng
+        img2 = psf2.drawImage(rng=rng, **drawKwargs)  # this is a copy of a withGSP
+        assert img2 == img
+
+    # And finally, look at copies of undrawn withGSP PSFs.
+    testimgs = []
+    for copier in copiers:
+        psf = psl.makePSF(exptime=0.02, time_step=0.01, diam=1.1, lam=1000.0)
+        psfGSP = psf.withGSParams(galsim.GSParams(folding_threshold=6e-3))  # hasn't been drawn yet
+        psf2 = copier(psfGSP)  # copy of undrawn
+        assert psf2 is not psfGSP
+        rng = galsim.BaseDeviate(57721) # reset rng
+        testimgs.append(psf2.drawImage(rng=rng, **drawKwargs))
+    # Draw the (fresh) withGSP PSF and compare
+    psf = psl.makePSF(exptime=0.02, time_step=0.01, diam=1.1, lam=1000.0)
+    psfGSP = psf.withGSParams(galsim.GSParams(folding_threshold=6e-3))
+    rng = galsim.BaseDeviate(57721)
+    img = psfGSP.drawImage(rng=rng, **drawKwargs)
+    for testimg in testimgs:
+        assert testimg == img
```

Comment:
I think we should maybe also do at least some of the above tests for OpticalPSF.  For instance, I think the following should probably fail currently:
```
opt_psf2 = opt_psf.withGSParams(gsp)
assert isinstance(opt_psf2, galsim.OpticalPSF)
```
which I think used to work. (?)  Also, I believe my suggested implementation of `withGSParams` should do this right, but I'm not 100% positive whether it will pass all of these other tests.

Reply:
I think this actually still works.  `OpticalPSF` isn't a subclass of `PhaseScreenPSF` (instead it holds one as a lazy property).  So `OpticalPSF.withGSParams` is really `GSObject.withGSParams`, which sets the `._gsparams` attr before the lazy property is evaluated.  I will certainly add this test though.

Reply:
Ah.  Ok.  I assumed it was, but you're right.